### PR TITLE
Documenting on-the-fly custom embedded document creation

### PR DESCRIPTION
**Pull request by brimoor**
Documents a previously-possible-but-undocumented pattern for creating custom embedded documents:
- declare an empty `DynamicEmbeddedDocument` field
- add dynamic attributes to it as desired

## Example usage

```py
import fiftyone as fo

dataset = fo.Dataset()

# Define an empty embedded document field
dataset.add_sample_field(
    "camera_info",
    fo.EmbeddedDocumentField,
    embedded_doc_type=fo.DynamicEmbeddedDocument,
)

# Declare a new `camera_id` attribute
dataset.add_sample_field("camera_info.camera_id", fo.StringField)
assert "camera_info.camera_id" in dataset.get_field_schema(flat=True)

# Includes a new `quality` attribute
sample1 = fo.Sample(
    filepath="/path/to/image1.jpg",
    camera_info=fo.DynamicEmbeddedDocument(
        camera_id="123456789",
        quality=51.0,
    ),
)

sample2 = fo.Sample(
    filepath="/path/to/image2.jpg",
    camera_info=fo.DynamicEmbeddedDocument(camera_id="123456789"),
)

dataset.add_samples([sample1, sample2], dynamic=True)
assert "camera_info.quality" in dataset.get_field_schema(flat=True)

# Populate a new `description` attribute
dataset.set_values("camera_info.description", ["foo", "bar"], dynamic=True)
assert "camera_info.description" in dataset.get_field_schema(flat=True)
```

### CODE REVIEW

File: fiftyone/core/odm/mixins.py [hunk 1]
```diff
@@ -1552,7 +1552,11 @@ def _add_field_doc(field_docs, field_or_doc):
             field_docs[i] = new_field_doc
             return
 
-    field_docs.append(new_field_doc)
+    try:
+        field_docs.append(new_field_doc)
+    except ReferenceError:
+        # mongoengine seems to lose references to things... it's okay
```
Comment (swheaton) [inline fiftyone/core/odm/mixins.py:1558]: uh what? are we sure this isn't hiding some other problem?

Comment (brimoor) [reply fiftyone/core/odm/mixins.py:1558]: It's hiding something, but it's some weird mongoengine garbage collection issue that I don't understand.

Sit down for this one... without the `try-except` I added, the following code works if you paste the whole thing into ipython at once, but it raises an error if you wait one second between the `add_sample_field()` calls:

```py
import fiftyone as fo

dataset = fo.Dataset()

dataset.add_sample_field(
    "camera_info",
    fo.EmbeddedDocumentField,
    embedded_doc_type=fo.DynamicEmbeddedDocument,
)

# This fails if you wait one second between running this and the above code
dataset.add_sample_field("camera_info.camera_id", fo.StringField)
```

Comment (brimoor) [reply fiftyone/core/odm/mixins.py:1558]: I'm confident that everything works as expected if we just ignore this error. The last thing that happens at the end of `add_sample_field()` is a call to `dataset._reload()`, which recreates the relevant mongoengine classes and should paint over any weird issues due to mongoengine's `append()` implementation failing.

The unit tests I added verify that the dataset functions as intended.

Comment (brimoor) [reply fiftyone/core/odm/mixins.py:1558]: Thanks for the deep dive!

**Pull request merged**